crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -0,0 +1,206 @@
+                }))
+                .build()?;
+
+            Arc::new(store)
+        } else {
+            object_store::parse_url_opts(&destination, &snapshots_options.additional_options)?
+                .0
+                .into()
+        };
+
+        let prefix = destination.path().into();
+        Ok(SnapshotRepository {
+            object_store,
+            destination,
+            prefix,
+            staging_path: base_dir.clone().join("snapshot-staging"),
+        })
+    }
+
+    /// Write a partition snapshot to the snapshot repository.
+    pub(crate) async fn put(
+        &self,
+        partition_id: PartitionId,
+        metadata: &PartitionSnapshotMetadata,
+        snapshot_path: &Path,
+    ) -> anyhow::Result<()> {
+        let snapshot_id = metadata.snapshot_id;
+        let lsn = metadata.min_applied_lsn;
+
+        debug!(
+            %snapshot_id,
+            partition_id = ?partition_id,
+            %lsn,
+            "Publishing partition snapshot to: {}",
+            self.destination,
+        );
+
+        // All common object stores list objects in lexicographical order, with no option for
+        // reverse order. We inject an explicit sort key into the snapshot prefix to make sure that
+        // the latest snapshot is always first.
+        let inverted_sort_key = format!("{:016x}", u64::MAX - lsn.as_u64());
+
+        // The snapshot data / metadata key format is: [<base_prefix>/]<partition_id>/<sort_key>_<lsn>_<snapshot_id>.tar
+        let snapshot_key = match self.prefix.as_str() {
+            "" | "/" => format!(
+                "{partition_id}/{sk}_{lsn}_{snapshot_id}.tar",
+                sk = inverted_sort_key,
+                lsn = metadata.min_applied_lsn,
+            ),
+            prefix => format!(
+                "{trimmed_prefix}/{partition_id}/{sk}_{lsn}_{snapshot_id}.tar",
+                trimmed_prefix = prefix.trim_start_matches('/').trim_end_matches('/'),
+                sk = inverted_sort_key,
+            ),
+        };
+
+        let staging_path = self.staging_path.clone();
+        tokio::fs::create_dir_all(&staging_path).await?;
+
+        let snapshot_path = snapshot_path.to_path_buf();
+        let packaging_task = task_center().spawn_blocking_unmanaged(
+            "package-snapshot",
+            Some(partition_id),
+            async move {
+                trace_span!("package-snapshot", %snapshot_id).in_scope(|| {
+                    let mut tarball = tar::Builder::new(NamedTempFile::new_in(&staging_path)?);
+                    debug!(
+                        "Creating snapshot tarball of {:?} in: {:?}...",
+                        &staging_path,
+                        tarball.get_ref()
+                    );
+                    tarball.append_dir_all(".", snapshot_path)?;
+                    tarball.finish()?;
+                    tarball.into_inner()
+                })
+            },
+        );
```

Comment:
Why are we creating a tar? Because it is simpler to download it? Because it saves some put/get requests (cost efficiency)? Because it makes the upload more efficient (performance)?

Reply:
It was a suggestion from both Jack and Ahmed that I went along with, believing I can stream the archive directly to S3. The upside is:

- simpler bucket layout with a single file (though that'll be short-lived as we move towards incremental snapshots)
- simpler to split up into chunks for multi-part upload

The major downside is that we have 2x the on-node storage footprint, AND we do a bunch of I/O to boot. I don't think it's pulling its weight, but I only just realized yesterday that I couldn't make streaming the tar work in memory (blocked on https://github.com/tokio-rs/tokio/issues/6914 which doesn't seem to be going anywhere).

I'll be removing the tar archiving in a follow-up iteration.